### PR TITLE
yanagiragi-hw3-attempt1

### PR DESCRIPTION
**Pull request by yanagiragi**
---
A good PR should include the following parts:

- [x] A clear title
- [x] A meaningful message for PR, as well as its commits
- [x] from your specific branch (***not master***) merging to teacher's `HW` branch (***not master***, again)
- [x] excluding any irrelevant files, such as binaries, text files, or dot files
- [x] passing tests/CI

CI is not passed, set PRELOAD_MKL should work (tested in AWS). However due to the handin policy I am not allowed to modify the validate.py to set PRELOAD_MKL.

 Let me know if there's a trick to set the variable without touching validate.py :D

Edit: The CI is now passed.

### CODE REVIEW

**Comment by yungyuc** [comment]
`validate.py` will set the environment variable `PRELOAD_MKL=1` if you have a file named `PRELOAD_MKL` in your directory: https://github.com/yungyuc/nsdhw_20sp/blob/master/hw3/validate.py#L10

**Comment by yanagiragi** [comment]
@yungyuc Your right! I must be silly working in midnight. Thanks a lot! 👍

File: hw3/yanagiragi/matrix.cpp [hunk 1]
```diff
@@ -0,0 +1,120 @@
...
+
+Matrix Matrix::multiply_mkl (const Matrix &left, const Matrix& right) 
+{
+    if (left.ncol() != right.nrow()) {
+        throw std::out_of_range("Error Size");
+    }
+    
+    Matrix ret(left.nrow(), right.ncol());
+
+    // Ref: https://software.intel.com/en-us/mkl-tutorial-c-multiplying-matrices-using-dgemm
```
Comment (tigercosmos) [inline hw3/yanagiragi/matrix.cpp:78]: cool!

File: hw3/yanagiragi/test.py [hunk 2]
```diff
@@ -0,0 +1,77 @@
...
+        a_mat = _matrix.Matrix(m, k, a_np.flatten(order='C')) # row major, C for C-style
+        b_mat = _matrix.Matrix(k, n, b_np.flatten(order='C'))
+        
+        start = time.time()
+        c_mat_naive = _matrix.multiply_naive(a_mat, b_mat)
+        end = time.time()
+        elapsed_naive += end - start
+
+        a_np = np.random.randn(m, k)
+        b_np = np.random.randn(k, n)
```
Comment (tigercosmos) [inline hw3/yanagiragi/test.py]: @yanagiragi  You use two different data for testing two functions.

The goal is to see the performance difference, we would like to decrease any factors that affect the result.
Therefore, you would like to create data sets first, and use the same data sets for testing.

Comment (tigercosmos) [reply hw3/yanagiragi/test.py]: you may need to fix this to get 5.

Comment (yanagiragi) [reply hw3/yanagiragi/test.py]: You were right. Will fix!

**Comment by tigercosmos** [comment]
get 5

**Pull request merged**